torch/nn/gradients.py

```diff
@@ -0,0 +1,317 @@
+"""Gradient interface"""
+
+import torch
+from .modules.utils import _single, _pair, _triple
+
+
+def input_grad_padding(out_backprop, input_size, stride, padding, kernel_size):
+    input_size = list(input_size)
+    k = out_backprop.dim() - 2
+
+    if len(input_size) == k + 2:
+        input_size = input_size[-k:]
+    if len(input_size) != k:
+        raise ValueError("input_size must have {} elements (got {})"
+                         .format(k + 2, len(input_size)))
+
+    def dim_size(d):
+        return ((out_backprop.size(d + 2) - 1) * stride[d] - 2 * padding[d] +
+                kernel_size[d])
+
+    min_sizes = [dim_size(d) for d in range(k)]
+    max_sizes = [min_sizes[d] + stride[d] - 1 for d in range(k)]
+    for size, min_size, max_size in zip(input_size, min_sizes, max_sizes):
+        if size < min_size or size > max_size:
+            raise ValueError(
+                ("requested an input grad size of {}, but valid sizes range "
+                 "from {} to {} (for an out_backprop of {})").format(
+                     input_size, min_sizes, max_sizes,
+                     out_backprop.size()[2:]))
+
+    return tuple(input_size[d] - min_sizes[d] for d in range(k))
+
+
+def conv1d_input(input_size, weight, out_backprop, stride=1, padding=0, dilation=1, groups=1, bias=None):
+    r"""
+    Computes the gradient of conv1d with respect to the input of the convolution.
+    This is same as the 1D transposed convolution operator under the hood but requires
+    the shape of the gradient w.r.t. input to be specified explicitly.
+
+    Args:
+        input_size : Shape of the input gradient tensor
```

Comment:
Wouldn't it be possible to require something simpler than the full input shape and infer some parameters? E.g. you can certainly do this with #channels, but the sizes might be ambiguous, right?

Reply:
Yes we can do that. We can calculate the batch and channels of input from weight and output shape.